Build failure: swt on staging-next

### Nixpkgs version

- Unstable (25.11)

### Steps to reproduce

git checkout staging-next
nix build .#swt
@doronbehar 

### Can Hydra reproduce this build failure?

Hydra doesn’t try to build the package.

### Link to Hydra build job

_No response_

### Relevant log output

```console
nix build .#swt            
error: Cannot build '/nix/store/jjnmm5pw28f6ldm3ln75pvnh4rs094jp-swt-4.34.drv'.
       Reason: builder failed with exit code 2.
       Output paths:
         /nix/store/ris1xirxwyk7c1rnr7q1cdyrijb2l9vs-swt-4.34
       Last 25 log lines:
       >  1541 |         rc = (jint)gdk_pixbuf_animation_iter_get_delay_time((GdkPixbufAnimationIter *)arg0);
       >       |         ^~
       > /nix/store/qpdga8b7a1in9l0lw90s1vmh5q0qnpaf-gdk-pixbuf-2.44.3-dev/include/gdk-pixbuf-2.0/gdk-pixbuf/gdk-pixbuf-animation.h:104:25: note: declared here
       >   104 | int                     gdk_pixbuf_animation_iter_get_delay_time             (GdkPixbufAnimationIter *iter);
       >       |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
       > os.c: In function 'Java_org_eclipse_swt_internal_gtk_GDK_gdk_1pixbuf_1animation_1iter_1get_1pixbuf':
       > os.c:1553:9: error: 'gdk_pixbuf_animation_iter_get_pixbuf' is deprecated [-Werror=deprecated-declarations]
       >  1553 |         rc = (jlong)gdk_pixbuf_animation_iter_get_pixbuf((GdkPixbufAnimationIter *)arg0);
       >       |         ^~
       > /nix/store/qpdga8b7a1in9l0lw90s1vmh5q0qnpaf-gdk-pixbuf-2.44.3-dev/include/gdk-pixbuf-2.0/gdk-pixbuf/gdk-pixbuf-animation.h:106:25: note: declared here
       >   106 | GdkPixbuf              *gdk_pixbuf_animation_iter_get_pixbuf                 (GdkPixbufAnimationIter *iter);
       >       |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
       > os.c: In function 'Java_org_eclipse_swt_internal_gtk_GDK_gdk_1pixbuf_1loader_1get_1animation':
       > os.c:1759:9: error: 'gdk_pixbuf_loader_get_animation' is deprecated [-Werror=deprecated-declarations]
       >  1759 |         rc = (jlong)gdk_pixbuf_loader_get_animation((GdkPixbufLoader *)arg0);
       >       |         ^~
       > In file included from /nix/store/qpdga8b7a1in9l0lw90s1vmh5q0qnpaf-gdk-pixbuf-2.44.3-dev/include/gdk-pixbuf-2.0/gdk-pixbuf/gdk-pixbuf.h:39:
       > /nix/store/qpdga8b7a1in9l0lw90s1vmh5q0qnpaf-gdk-pixbuf-2.44.3-dev/include/gdk-pixbuf-2.0/gdk-pixbuf/gdk-pixbuf-loader.h:102:22: note: declared here
       >   102 | GdkPixbufAnimation * gdk_pixbuf_loader_get_animation (GdkPixbufLoader *loader);
       >       |                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
       > cc1: all warnings being treated as errors
       > make[1]: *** [make_linux.mak:144: os.o] Error 1
       > make[1]: Leaving directory '/build/source/library'
       > *** GTK3 Build failed, aborting further actions..
       > make: *** [makefile:16: gtk3] Error 2
       For full logs, run:
         nix log /nix/store/jjnmm5pw28f6ldm3ln75pvnh4rs094jp-swt-4.34.drv
```

### Additional context

swt is being modified in this pr https://github.com/NixOS/nixpkgs/pull/458083
### System metadata

nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.17.7-cachyos, NixOS, 25.11 (Xantusia), 25.11.20251106.f7213c5`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.31.2`
 - nixpkgs: `/nix/store/1rqn7pcwhk0dqaydd0a024y3gclxp7zg-source`



### Notify maintainers



---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/sw/swt/package.nix
+++ b/pkgs/by-name/sw/swt/package.nix
@@ -31,7 +31,7 @@ stdenv.mkDerivation (finalAttrs: {
     x86_64-darwin.platform = "cocoa-macosx-x86_64";
     x86_64-darwin.hash = "sha256-Uns3fMoetbZAIrL/N0eVd42/3uygXakDdxpaxf5SWDI=";
     aarch64-darwin.platform = "cocoa-macosx-aarch64";
-    aarch64-darwin.hash = "sha256-Uns3fMoetbZAIrL/N0eVd42/3uygXakDdxpaxf5SWDI";
+    aarch64-darwin.hash = "sha256-jvxmoRFGquYClPgMqWi2ylw26YiGSG5bONnM1PcjlTM=";
   };
   passthru.srcMetadata =
     finalAttrs.passthru.srcMetadataByPlatform.${stdenv.hostPlatform.system} or null;
@@ -47,25 +47,29 @@ stdenv.mkDerivation (finalAttrs: {
       url = "https://archive.eclipse.org/eclipse/downloads/drops4/R-${finalAttrs.fullVersion}/swt-${finalAttrs.version}-${srcMetadata.platform}.zip";
       inherit (srcMetadata) hash;
       stripRoot = false;
-      postFetch = ''
-        mkdir "$unpackDir"
-        cd "$unpackDir"
+      postFetch =
+        # On Linux, extract and use only the sources from src.zip
+        lib.optionalString stdenv.hostPlatform.isLinux ''
+          mkdir "$unpackDir"
+          cd "$unpackDir"
 
-        renamed="$TMPDIR/src.zip"
-        mv -- "$out/src.zip" "$renamed"
-        unpackFile "$renamed"
-        rm -r -- "$out"
+          renamed="$TMPDIR/src.zip"
+          mv -- "$out/src.zip" "$renamed"
+          unpackFile "$renamed"
+          rm -r -- "$out"
 
-        mv -- "$unpackDir" "$out"
-      '';
+          mv -- "$unpackDir" "$out"
+        '';
     };
 
   nativeBuildInputs = [
     jdk
     stripJavaArchivesHook
+  ]
+  ++ lib.optionals stdenv.hostPlatform.isLinux [
     pkg-config
   ];
-  buildInputs = [
+  buildInputs = lib.optionals stdenv.hostPlatform.isLinux [
     gtk3
     libGLU
   ];
@@ -76,14 +80,17 @@ stdenv.mkDerivation (finalAttrs: {
   OUTPUT_DIR = "${placeholder "out"}/lib";
   # GTK4 is not supported yet. See:
   # https://github.com/eclipse-platform/eclipse.platform.swt/issues/652
-  makeFlags = "gtk3";
-  preBuild = ''
+  makeFlags = lib.optionals stdenv.hostPlatform.isLinux [ "gtk3" ];
+
+  NIX_CFLAGS_COMPILE = lib.optionalString stdenv.hostPlatform.isLinux "-std=gnu17";
+  postPatch = lib.optionalString stdenv.hostPlatform.isLinux "substituteInPlace library/make_linux.mak --replace-fail 'CFLAGS += -Werror' ''";
+  preBuild = lib.optionalString stdenv.hostPlatform.isLinux ''
     cd library
     mkdir -p ${finalAttrs.OUTPUT_DIR}
   '';
 
-  # Build the jar
-  postBuild = ''
+  # Build the jar (Linux only, Darwin uses prebuilt)
+  postBuild = lib.optionalString stdenv.hostPlatform.isLinux ''
     cd ../
     mkdir out
     find org/ -name '*.java' -type f -exec javac -encoding utf8 -d out/ {} +
@@ -97,9 +104,27 @@ stdenv.mkDerivation (finalAttrs: {
     runHook preInstall
 
     install -d -- "$out/jars"
+
+  ''
+  # On Darwin, use the prebuilt swt.jar which includes native libraries
+  + lib.optionalString stdenv.hostPlatform.isDarwin ''
+    # Remove signature files to avoid validation errors after stripJavaArchivesHook modifies the jar
+    mkdir -p jar-temp
+    cd jar-temp
+    ${jdk}/bin/jar -xf ../swt.jar
+    rm -f META-INF/*.SF META-INF/*.RSA META-INF/*.DSA
+    ${jdk}/bin/jar -cf "$out/jars/swt.jar" *
+    cd ..
+    rm -rf jar-temp
+
+  ''
+  # On Linux, build from source
+  + lib.optionalString stdenv.hostPlatform.isLinux ''
     install -m 644 -t out -- version.txt
     (cd out && jar -c *) > "$out/jars/swt.jar"
 
+  ''
+  + ''
     runHook postInstall
   '';
 

